trulens_eval/trulens_eval/feedback/v2/provider/base.py

```diff
@@ -0,0 +1,95 @@
+from abc import abstractmethod
+from typing import Iterable, Optional
+
+from trulens_eval.feedback.provider.endpoint.base import Endpoint
+from trulens_eval.trulens_eval.feedback.base import ClassificationModel, Hate, HateThreatening, Model, WithExamples, WithPrompt
+from trulens_eval.utils.pyschema import WithClassInfo
+from trulens_eval.utils.serial import SerialModel
+
+# Level 4 feedback abstraction
+
+class Provider(SerialModel, WithClassInfo):
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    endpoint: Optional[Endpoint]
+
+    def __init__(self, name: str = None, **kwargs):
+        # for WithClassInfo:
+        kwargs['obj'] = self
+
+        super().__init__(*args, **kwargs)
+
+    @abstractmethod
+    def supported_models(self) -> Iterable[Model]:
+        pass
+
+    @abstractmethod
+    def complete(self, model: Model) -> str:
+        # OpenAI completion models
+        # Cohere completion models
```

Comment:
Add LiteLLM completion models here too.

Reply:
Deeper level work is ongoing. Lets skip this for subsequent PRs.